### ISSUE TITLE
Advanced search options

### ISSUE THREAD

**Issue by CJ-Wright** [intake/intake#426]
Should there be more advanced search options? Would it be possible to incorporate something like [mongoquery](https://github.com/kapouille/mongoquery)? Or use elasticsearch for the metadata?

**Comment by martindurant** [comment]
There is already some discussion on this, particularly about exposing the set of entries as a dataframe, so people can manipulate using any pandas flow. I would *not* want to externalise to another system like mongo/ES, but we could incorporate more sophisticated things that map dicts or do semantic text stuff. The field is wide open...

**Comment by CJ-Wright** [comment]
`mongoquery` might work for this, it applies `mongo` like syntax to lists of python dictionaries. This would enable more advanced queries than "is word anywhere in dict".

How would you normalize the entries into a dataframe? My impression was that the entry metadata stored in the `foo.yaml` files were essentially a document store.

**Comment by martindurant** [comment]
> How would you normalize the entries into a dataframe?

Indeed, this was only discussed in the context of specialist catalogues, where all of the entries are guaranteed to be homogenous.

**Comment by martindurant** [comment]
> it applies mongo like syntax to lists of python dictionaries

I've never used it, so perhaps it would be nice to include some examples. I assume it would be easy to offer as an alternative, if it happens to be installed.

**Comment by CJ-Wright** [comment]
We used mongoquery to act like mongo in front of the portable databrokers at the NSLS-II.

Would it be worthwhile to add a `key` keyword to the search method? Each entry would be passed into the key function and added to the new cat. if the function returns true.

**Comment by martindurant** [comment]
sure, that's a fine way to do it - `key` could be "text", the current and default behaviour, "mongoq" for your suggestion (if installed) or any arbitrary callable. The callable takes entries as input, bool output?

**Comment by CJ-Wright** [comment]
> The callable takes entries as input, bool output

Yep. You could even implement fuzzy searching with string distances.
